operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go

```diff
@@ -0,0 +1,253 @@
+			if template.Spec.EnvironmentList[i].Name == isnap.Spec.Environment.Name {
+				env = &template.Spec.EnvironmentList[i]
+				break
+			}
+		}
+
+		// Check if the specified environment was found.
+		if env == nil {
+			return false, fmt.Errorf("environment %s not found in template %s. It is not possible to complete the InstanceSnapshot %s",
+				isnap.Spec.Environment.Name, template.Name, isnap.Name)
+		}
+	} else {
+		// If the environment is not explicitly declared, take the first one.
+		env = &template.Spec.EnvironmentList[0]
+	}
+
+	// Check if the environment is a persistent VM.
+	if env.EnvironmentType != crownlabsv1alpha2.ClassVM || !env.Persistent {
+		return false, fmt.Errorf("environment %s is not a persistent VM. It is not possible to complete the InstanceSnapshot %s",
+			env.Name, isnap.Name)
+	}
+
+	// Check if the VM is running.
+	if instance.Spec.Running || instance.Status.Phase != "VmiOff" {
+		return false, fmt.Errorf("the vm is running. It is not possible to complete the InstanceSnapshot %s", isnap.Name)
+	}
+
+	return false, nil
+}
+
+// GetJobStatus sets a Job and returns its status.
+func (r *InstanceSnapshotReconciler) GetJobStatus(job *batch.Job) (bool, batch.JobConditionType) {
+	for _, c := range job.Status.Conditions {
+		// If the status corresponding to Success or failed is true, it means that the job completed.
+		if c.Status == corev1.ConditionTrue && (c.Type == batch.JobFailed || c.Type == batch.JobComplete) {
+			return true, c.Type
+		}
+	}
+
+	// Job did not complete.
+	return false, ""
+}
+
+// GetSnapshottingJob generates the job to be created.
+func (r *InstanceSnapshotReconciler) GetSnapshottingJob(isnap *crownlabsv1alpha2.InstanceSnapshot, job *batch.Job) {
+	var backoff int32 = 2
+	imagedir := isnap.Spec.Instance.Namespace
+	imagetag := fmt.Sprint(time.Now().Format("02012006t150405"))
+	volumename := isnap.Spec.Instance.Name
+
+	// Define volumes.
+
+	// Define VM VolumeSource.
+	persistentvolsrc := corev1.PersistentVolumeClaimVolumeSource{
+		ClaimName: volumename,
+	}
+	vmvolume := corev1.VolumeSource{
+		PersistentVolumeClaim: &persistentvolsrc,
+	}
+
+	// Define temp VolumeSource.
+	emptyvolsrc := corev1.EmptyDirVolumeSource{}
+	tmpvol := corev1.VolumeSource{
+		EmptyDir: &emptyvolsrc,
+	}
+
+	// Define secret VolumeSource.
+	secretvolsrc := corev1.SecretVolumeSource{
+		SecretName: "kaniko-secret",
+		Items: []corev1.KeyToPath{
+			{
+				Key:  ".dockerconfigjson",
+				Path: "config.json",
+			},
+		},
+	}
+	secretvol := corev1.VolumeSource{
+		Secret: &secretvolsrc,
+	}
+
+	volumes := []corev1.Volume{
+		{
+			Name:         volumename,
+			VolumeSource: vmvolume,
+		},
+		{
+			Name:         "tmp-vol",
+			VolumeSource: tmpvol,
+		},
+		{
+			Name:         "kaniko-secret",
+			VolumeSource: secretvol,
+		},
+	}
+
+	// Define containers.
+
+	// Define Docker pusher container.
+	pushcontainer := corev1.Container{
+		Name:  "docker-pusher",
+		Image: "gcr.io/kaniko-project/executor:latest",
```

Comment:
Never use latest in production, as it is a recipe for disaster. The best would be to add a new flag to the operator to configure it